### ISSUE TITLE
Documentation on communication protocol with lizzie

### ISSUE THREAD

**Issue by infinity0** [gcp/leela-zero#2053]
lizzie has a very [ad-hoc way](https://github.com/featurecat/lizzie/blob/6fa818efbae2f730d9e370db3c91c67b86af544d/src/main/java/featurecat/lizzie/analysis/Leelaz.java#L220) of parsing the output from leelaz, leading to issues like https://github.com/featurecat/lizzie/issues/455 - it would be good for leelaz to document the syntax somewhere.

**Comment by roy7** [comment]
If it helps, the commands were added in https://github.com/gcp/leela-zero/pull/1388

**Comment by infinity0** [comment]
Looks like there is other output as well as what was added in that PR, and that's what's confusing lizzie.

**Comment by gcp** [comment]
That output is on stderr and not part of GTP. So documenting the GTP stuff won't do anything for UIs that are trying to parse the stderr output (which isn't standardized and probably shouldn't be!).

Rather, we should see if there's good way to send lizzie the info it needs to know in a format that is compatible with GTP.

**Comment by tommadams** [comment]
Hello, author of Minigo's UI here!

I thought it might be worth chiming in with a description of our approach, which has been working pretty well for us.

I'm not familiar with how Lizzie works but I suspect the Minigo UI is somewhat similar in that it talks to the backend via GTP, and the backend reports data that the UI needs over `stderr`.

We chose to write to `stderr` because the GTP spec limits what can be written to `stdout`. To take the `genmove` command as an example: we wanted the UI to show a live update of the variations being explored during tree search while handling the command, but the GTP spec states that the only thing that can be printed to `stdout` is the vertex of the move played move, `"pass"` or `"resign"`.

As for the data we send...

The Minigo UI doesn't understand the rules of go and is completely reliant on the backend to provide all the data it needs to function.

In order to avoid ad-hoc parsing in the UI code, the Minigo backend writes the data that the UI needs as JSON objects. Each object is written on a separate line, prefixed with a known string (e.g. `"mg-position:"`, `"mg-update:"`) so that the UI can easily distinguish the JSON data from other messages that get written to `stderr`.

The backend currently writes two kinds of objects:
 - [`PositionJson`](https://github.com/tensorflow/minigo/blob/a840cb5d05cee0b3358d86d183ec3f709ce0a452/minigui/app.ts#L21) contains a description of the board and game state. It is written after every move played.
 - [`PositionUpdateJson`](https://github.com/tensorflow/minigo/blob/a840cb5d05cee0b3358d86d183ec3f709ce0a452/minigui/app.ts#L33) constains an incremental update to the current board position, e.g. number of reads, tree search's current principle variation, tree search's most recently read variation, etc. This is written every ~50ms during tree search.

I think there may be some benefit to standardizing on a common protocol for sending data to UIs and external tools. If modern go engines shared a common protocol, it would facilitate easier sharing of tools, and allow two different engines to play in the same UI.

**Comment by yishn** [comment]
I feel like Leela's [analyze commands](https://github.com/gcp/leela-zero/pull/1388) are superior, since it's GTP compliant (meaning GTP controllers which don't understand them play just fine) and not relying on stderr which may potentially contain any kind of free form data. 

FYI the syntax has undergone some changes sind the original pull request.

**Comment by tommadams** [comment]
@roy7 & @yishn Thanks a lot for the references! The discussion in those pull requests is remarkably similar to the conversations we had.

We also wrestled with the same choices between writing to `stdout` or `stderr` and whether or not to have a custom command for performing genmove while reporting the dynamic search state. We even implemented a custom `mg-genmove` command that was very similar to `lz-genmove_analyze` without having seen your code, so I guess it must be a good idea :)

In the end, we abandoned `mg-genmove` and decided that using the standard GTP `genmove` command and writing to `stderr` was preferable, but I certainly understand why others may think differently.

We shared the concerns that `stderr` may contain free form data but it never became a problem in practice and using a unique line prefix has proven robust. If you add a newline to the start of the prefix, you avoid the problem of other code writing to `stderr` without a trailing newline.

@yishn I'm afraid that I don't understand your point about Leela's solution being GTP compliant. Since the `lz-genmove_analyze` isn't a standard GTP command, doesn't that mean analysis won't work with a standard GTP controller? Sorry if this is a basic question, I don't actually have much practical experience with GTP, which @sethtroisi can attest to.

**Comment by yishn** [comment]
>  I'm afraid that I don't understand your point about Leela's solution being GTP compliant. Since the lz-genmove_analyze isn't a standard GTP command, doesn't that mean analysis won't work with a standard GTP controller?

I just wanted to point out that we didn't change the behavior of the original `genmove` command, so engines can still use it. Also, that the GTP extension we have is also based on GTP, and not something entirely different.

But maybe I'm also a little bit opinionated here, since Sabaki already has the analyze commands implemented and I kinda don't want to change them... 😅

**Comment by tommadams** [comment]
Gotcha, thanks for the explanation.

We wanted to use the standard GTP commands without breaking compatibility with GTP controllers, which means we're forced to use `stderr` as far as I can tell.

**Comment by yishn** [comment]
We're not breaking compatibility with GTP controllers either, as I tried to explain.

**Comment by tommadams** [comment]
Sorry, I understand that. I meant to say that we wanted dynamic reporting of the search state to still work even when we are playing a game via a GTP controller. For that, we can't use a custom `genmove` command.

**Comment by yishn** [comment]
I'm afraid that I don't understand your point about dynamic reporting working with GTP controllers. If a GTP controller doesn't know how to parse your special stderr directives, there will be no dynamic reporting involved, no? Only a controller which knows how to handle the extra information can utilize them; it's the same as Leela's analyze commands in that regard: only a controller which support the analyze commands can utilize dynamic reporting.

There's another good thing about using pure GTP commands: You don't run into synchronization problems such as trying to figure out which update on stderr corresponds to which GTP board state. I think you solved that by having to provide an extra `id` field on both `PositionJson` and `PositionUpdateJson`.

**Comment by tommadams** [comment]
There are two use cases that I think are interesting in which I believe it makes sense to use the standard GTP commands rather than extensions.

1. Connecting the _UI_ (not the engine) to an existing GTP controller so the user can watch their engine thinking in real time as it's playing a game. The UI acts as an intermediary between the controller and the engine, forwarding commands from the controller to the engine and results from the engine back to the controller. While doing this, the UI monitors `stderr` to show the live search status. Since the GTP controller doesn't know about the dynamic reporting extensions, the engine, and by proxy the UI need to use the standard commands.

2. Writing a new GTP controller that supports standardized dynamic reporting GTP extensions. Imagine having a server running a GTP controller that allows two engines to play each other while showing a dynamically updating view of what each engine was thinking. I think that would be a compelling environment to watch computer Go games. In order to give the widest possible support, the GTP controller server should also work with engines that don't support dynamic reporting (it just wouldn't be able to show anything while that engine was thinking), which requires that it uses the standard GTP commands.

**Comment by roy7** [comment]
That reminds me of how the interface program for CGOS has a way to attach gogui so you can watch the games play out live with a real front end program. :)

**Comment by amj** [comment]
@tommadams just to clarify, in points 1 and 2, "standard GTP commands" mean basically only `genmove` and `play`, correct?

**Comment by tommadams** [comment]
@amj correct, thanks for clarifying.

**Comment by yishn** [comment]
Regarding 2: Sabaki (as a GTP controller) can handle both normal engines and engines with analyze commands just fine by conducting a simple feature detection first (checking whether the engine supports the `genmove_analyze` command using `list_commands`). This is similar to how proressive enhancement works.

IMHO it's better to have the UI be aware whether dynamic reporting is supported or not and when to expect them. By hiding dynamic reporting inside stderr, there's no clear way to determine.

I guess point 1 is a real disadvantage. Using the stderr channel, UIs can connect to legacy GTP controllers and still support dynamic reporting. However, I don't find making the distinction between UI and GTP controller that useful in the first place. The UI might as well be a controller.

**Comment by amj** [comment]
"the UI might as well be a controller" -- but it often isn't; the point of the controller is that it is ubiquitous and general.  How cool would it be to e.g. attach to a running test in ringmaster, or be able to spectate games on CGOS and get outputs from the engine playing?  @roy7 could host or stream the UI attached to his LZ bots as they play on OGS, for example.  There's no good way to do this if you rely on 'non-standard' GTP commands because the controller & the UI need to be decoupled.

**Comment by tommadams** [comment]
I see that the UI could translate the controllers commands (e.g. `genmove` -> `genmove_analyze`) before passing them on to the engine, if that's what you mean.

One thing I'm curious about in your scheme: how does the engine send dynamic reports about its search while it's pondering (waiting for the other player to make a move)? There's no GTP command for it to respond to.

Pondering aside, another feature that using to `stderr` made easy: Minigo performs iterative refinement of the estimated win rate for every position of every variation in a game after loading an SGF. It does this in the background in parallel with normal tree search, streaming the results over.

**Comment by yishn** [comment]
> One thing I'm curious about in your scheme: how does the engine send dynamic reports about its search while it's pondering (waiting for the other player to make a move)? There's no GTP command for it to respond to.

For that purpose we have the `lz-analyze` command which continuously outputs the info lines in its response until stdin gets more input.

> Pondering aside, another feature that using to stderr made easy: Minigo performs iterative refinement of the estimated win rate for every position of every variation in a game after loading an SGF. It does this in the background in parallel with normal tree search, streaming the results over.

This sounds extremely cool.

**Comment by roy7** [comment]
@yishn Just some love for Sabaki + Leela Zero I wanted to share with you. ;) https://forums.online-go.com/t/leela-zero-sabaki-had-really-transformed-into-a-fantastic-analyzing-tool/19307
